File: services/marketplace-service/src/main/java/com/designer/marketplace/lms/entity/Course.java
package com.designer.marketplace.lms.entity;

import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

/**
 * Course entity for Learning Management System.
 * Stored in MongoDB for flexible schema and nested content.
 */
@Document(collection = "courses")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EqualsAndHashCode
@ToString
public class Course {

    @Id
    private String id;

    @Indexed
    private String title;

    private String description;

    private String shortDescription;

    @Indexed
    private Long instructorId;  // References PostgreSQL user

    private String instructorName;

    private String thumbnailUrl;

    private String previewVideoUrl;

    @Indexed
    private CourseCategory category;

    @Indexed
    private CourseLevel level;

    @Indexed
    private CourseStatus status;

    private BigDecimal price;

    private String currency;

    @Builder.Default
    private List<String> tags = new ArrayList<>();

    @Builder.Default
    private List<String> objectives = new ArrayList<>();

    @Builder.Default
    private List<String> requirements = new ArrayList<>();

    @Builder.Default
    private List<Module> modules = new ArrayList<>();

    // Statistics
    @Builder.Default
    private Integer totalEnrollments = 0;

    @Builder.Default
    private Double averageRating = 0.0;

    @Builder.Default
    private Integer reviewCount = 0;

    @Builder.Default
    private Integer totalDurationMinutes = 0;

    @Builder.Default
    private Integer totalLessons = 0;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    private Instant publishedAt;

    // SEO
    private String slug;

    private String metaDescription;

    /**
     * Course status enum
     */
    public enum CourseStatus {
        DRAFT,
        PENDING_REVIEW,
        PUBLISHED,
        ARCHIVED
    }

    /**
     * Course level enum
     */
    public enum CourseLevel {
        BEGINNER,
        INTERMEDIATE,
        ADVANCED,
        ALL_LEVELS
    }

    /**
     * Course category enum
     */
    public enum CourseCategory {
        UI_DESIGN,
        UX_DESIGN,
        GRAPHIC_DESIGN,
        WEB_DEVELOPMENT,
        MOBILE_DEVELOPMENT,
        BRANDING,
        ILLUSTRATION,
        MOTION_GRAPHICS,
        PHOTOGRAPHY,
        VIDEO_EDITING,
        MARKETING,
        BUSINESS,
        OTHER
    }

    /**
     * Calculate and update total duration and lesson count from modules
     */
    public void recalculateStats() {
        int totalMins = 0;
        int lessonsCount = 0;
        
        if (modules != null) {
            for (Module module : modules) {
                if (module.getLessons() != null) {
                    lessonsCount += module.getLessons().size();
                    for (Lesson lesson : module.getLessons()) {
                        if (lesson.getDurationMinutes() != null) {
                            totalMins += lesson.getDurationMinutes();
                        }
                    }
                }
            }
        }
        
        this.totalDurationMinutes = totalMins;
        this.totalLessons = lessonsCount;
    }
}

File: services/lms-service/Models/Course.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace LmsService.Models;

public class Course
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = null!;

    [BsonElement("title")]
    public string Title { get; set; } = null!;

    [BsonElement("description")]
    public string Description { get; set; } = string.Empty;

    [BsonElement("shortDescription")]
    public string ShortDescription { get; set; } = string.Empty;

    [BsonElement("instructorId")]
    public long InstructorId { get; set; }

    [BsonElement("instructorName")]
    public string InstructorName { get; set; } = string.Empty;

    [BsonElement("thumbnailUrl")]
    public string? ThumbnailUrl { get; set; }

    [BsonElement("previewVideoUrl")]
    public string? PreviewVideoUrl { get; set; }

    [BsonElement("category")]
    public CourseCategory Category { get; set; }

    [BsonElement("level")]
    public CourseLevel Level { get; set; }

    [BsonElement("status")]
    public CourseStatus Status { get; set; }

    [BsonElement("price")]
    public decimal Price { get; set; }

    [BsonElement("currency")]
    public string Currency { get; set; } = "USD";

    [BsonElement("tags")]
    public List<string> Tags { get; set; } = new();

    [BsonElement("objectives")]
    public List<string> Objectives { get; set; } = new();

    [BsonElement("requirements")]
    public List<string> Requirements { get; set; } = new();

    [BsonElement("modules")]
    public List<Module> Modules { get; set; } = new();

    [BsonElement("totalEnrollments")]
    public int TotalEnrollments { get; set; }

    [BsonElement("averageRating")]
    public double AverageRating { get; set; }

    [BsonElement("reviewCount")]
    public int ReviewCount { get; set; }

    [BsonElement("totalDurationMinutes")]
    public int TotalDurationMinutes { get; set; }

    [BsonElement("totalLessons")]
    public int TotalLessons { get; set; }

    [BsonElement("slug")]
    public string Slug { get; set; } = string.Empty;

    [BsonElement("metaDescription")]
    public string? MetaDescription { get; set; }

    [BsonElement("createdAt")]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [BsonElement("updatedAt")]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    [BsonElement("publishedAt")]
    public DateTime? PublishedAt { get; set; }

    public void RecalculateStats()
    {
        int totalMins = 0;
        int lessonsCount = 0;

        foreach (var module in Modules)
        {
            foreach (var lesson in module.Lessons)
            {
                totalMins += lesson.DurationMinutes;
                lessonsCount++;
            }
        }

        TotalDurationMinutes = totalMins;
        TotalLessons = lessonsCount;
    }
}

public class Module
{
    [BsonElement("id")]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [BsonElement("title")]
    public string Title { get; set; } = null!;

    [BsonElement("description")]
    public string? Description { get; set; }

    [BsonElement("orderIndex")]
    public int OrderIndex { get; set; }

    [BsonElement("lessons")]
    public List<Lesson> Lessons { get; set; } = new();
}

public class Lesson
{
    [BsonElement("id")]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [BsonElement("title")]
    public string Title { get; set; } = null!;

    [BsonElement("description")]
    public string? Description { get; set; }

    [BsonElement("type")]
    public LessonType Type { get; set; }

    [BsonElement("videoUrl")]
    public string? VideoUrl { get; set; }

    [BsonElement("videoKey")]
    public string? VideoKey { get; set; }

    [BsonElement("content")]
    public string? Content { get; set; }

    [BsonElement("durationMinutes")]
    public int DurationMinutes { get; set; }

    [BsonElement("orderIndex")]
    public int OrderIndex { get; set; }

    [BsonElement("isFree")]
    public bool IsFree { get; set; }

    [BsonElement("resources")]
    public List<LessonResource> Resources { get; set; } = new();
}

public class LessonResource
{
    [BsonElement("name")]
    public string Name { get; set; } = null!;

    [BsonElement("url")]
    public string Url { get; set; } = null!;

    [BsonElement("type")]
    public string Type { get; set; } = null!;
}

public enum CourseCategory
{
    UiDesign,
    UxDesign,
    GraphicDesign,
    WebDevelopment,
    MobileDevelopment,
    Branding,
    Illustration,
    MotionGraphics,
    Photography,
    VideoEditing,
    Marketing,
    Business,
    Other
}

public enum CourseLevel
{
    Beginner,
    Intermediate,
    Advanced,
    AllLevels
}

public enum CourseStatus
{
    Draft,
    PendingReview,
    Published,
    Archived
}

public enum LessonType
{
    Video,
    Text,
    Quiz,
    Assignment
}
